### ISSUE TITLE
UK Race Formats (Updated) - Feature Request/Enhancement

### ISSUE THREAD

**Issue by CerberusVelvet** [RotorHazard/RotorHazard#598]
Currently the RotorHazard system cannot be used to run a British Qualifier Event as per the British Drone Racing Association race format rules. Resulting in UK Race Directors being forced to pay for and use LiveTime software, which is not best suited to the RotorHazard timing hardware. This results in UK Race Directors purchasing ImmersionRC LapRF8 timers and LiveTime subscriptions to run BQE events.

To solve this I would like to propose changes/additions to the current race format options in RotorHazard to enable a RotorHazard solution to run a BQE race. 

The new (for 2022 but being trialed in 2021) BQE race format is as follows:-
Each heat is run once (or possibly more, depending on the situation). The race director then uses this race data to set a cap on the number of laps possible, without destroying a lipo. The 'practice' heats data is then discarded.
The heats race format is then set as "First to <X> laps". With a total race time decided by the Race Director, usually 2m with 30s 'overtime'.

For example. 
The Race Director, after practice, sets a lap limit of 6 laps. With a race time of 2m30s. The pilots perform as follows-
Pilot A - completes 6 laps in 1m50s 
Pilot B - completes 6 laps in 1m55s
Pilot C - completes 4 laps in 2m and then goes on to complete a 5th lap in 20s of overtime elapsed
Pilot D - crashes after 1m30s on their 4th lap and does not return to the air
Pilot E - completes 1 lap in 2m5s and is still half way through their second lap at the end of the 30s overtime
Pilot F - completes 5 laps in 2m6s and goes on to complete a 6th lap in 25th second of overtime

In this race-
The winner is declared as Pilot A after 1m50s. No more laps are counted for Pilot A. Pilot A flies to the landing zone and lands.
Pilot B finishes in second place after 1m55s. After crossing the line for their 6th lap no more laps are counted for Pilot B. Pilot B flies to the landing zone and lands.
Pilot C - Continues to fly after the 2m timer expires as they have not reached the lap cap. Pilot C goes on to complete their 5th lap before overtime expires. Pilot C finishes in 4th place recording 5 laps
Pilot D - Does not 'complete' the race in the air. But, having completed 4 laps is placed in 5th position
Pilot E - Continues to fly after the 2m timer but does not complete their 2nd lap in overtime. Having completed a total of 1 lap Pilot E finishes in 6th place and flies from their location when the overtime elapses, directly to the landing zone. 
Pilot F - Completes their 6th lap during overtime and is placed in 3rd

It is important that when designing a track the Race Director ensures the landing pad is placed in such a way that the race track is not overflown by any pilot traversing from the finish gate to the landing pad. This is not a consideration for RotorHazard. However it would remove uncertainty if once a pilots race is complete no more laps are counted for that pilot, even if the timer has not elapsed.

It is important to note, Once the winner is declared the race is not complete. The winner can be announdced (by RH callout) but the timer continues. It would be a good feature if the following pilots times and finishing places were called out as they cross the line. EG. "Pilot C finished, Position 2, 1m55s"

For pilot info there should be a callout at the 2 minute mark (in this example) "Time elapsed, finish your laps". Followed by, at the 2m30s mark (in this example) "The race is completed, land your quads"

Once the overtime timer has elapsed the race state should automatically move to 'stopped' and no more laps should be counted. This prevents any spurious laps being registered by a drone that is halfway through a lap when the timer expires, flying across the start/finish gate on the way to the landing pad. Whilst the 'stop' button is still worthwhile to remain in the case of a Race Director having to 'Red Flag' a race. It should not be needed to press 'stop' at the end of the overtime timer as the race has reached a defined stopping point and the race state should automatically be moved to 'stopped'.

It may be a worthwhile feature to allow the Race Director to select the RH behavior once the race completes. Choosing a setting in the UI to determine if the RH automatically saves the laps and moves to the next heat OR if the RH pauses and allows the Race Director to choose to save or discard the laps before the RH moves to the next heat.

**Comment by HazardCreative** [comment]
Can you please clarify what specifically requires the use of another product?

**Comment by CerberusVelvet** [comment]
> Can you please clarify what specifically requires the use of another product?

RotorHazard cannot, yet, run the required ‘race format’. Therefore UK Race Directors instead use LiveTime. Having canvassed some Race Directors, there are other LiveTime features such as scheduling and event management that also seem to sway this decision. But that’s probably going beyond the scope of this issue. The component of concern here is the race format. Which causes the RH to be untenable for a BQE. Despite its clearly superior pilot tracking. 

There is a perception amongst UK race organisers that a LapRF8 connected to LiveTime performs better than a RotorHazard timer connected to LiveTime. Thus UK race directors will currently not use RH to run a BQE. In fact, to date, a RH timer has never been used to run a BQE anywhere in the UK. This perception of RH is open to debate and there is only anecdotal evidence to support it. However several UK race directors, when asked, felt that it was inadvisable to attempt to run LT with RH. I don't have access to a LapRF8 to attempt any real world testing to prove or disprove the prevailing opinion. 

This, in the UK to run an official qualifier event other products are used. 

Hope that answers it. Let me know if not :)

**Comment by CerberusVelvet** [comment]
Don’t really understand your combative stance on this? 

If you don’t like the way the BDRA run BQE races in the UK. Well that’s between you and them. I didn’t choose the format. I just tried to explain it. It’s the hand I’ve been dealt as a UK race director. 

If you don’t wish RH to be able to run a BQE then I’ll respect your decision. Even if I don’t understand it. 

If RH can already run this format as simply as LiveTime then I’m open to someone showing me how and I will run a BQE, or at least a simulation of one, to show it can be done and help with the take up of RH as the premium choice for a BQE. As yet I’ve not found a way to run the BQE format as easily in RH as it can be in LiveTime. 

I’ve just tried to explain why RH isn’t used and how I feel it could be able to be used, with some alterations. 

I’m not debating which is better or which should or should not be used. I’m just trying to express how, as a user of RH and actually a vocal opponent of LiveTime, how RH could include a race format option that allows me to stop giving LiveTime my money and to use a far cleaner and more reliable timer for UK events.

Again, really not sure what all the push-back is about. I didn’t invent this race format. I just have to try to work with it. #everyoneshootsthemessenger

**Comment by CerberusVelvet** [comment]
I believe the UK ‘overtime’ and USA ‘overtime’ May define different concepts. 

The UK 30s ‘overtime’ is to allow the pilot to finish the lap they are on. Once they complete the lap they are on, they have finished their race. Similar to F1 qualifying where if a car crossed the line just before the flag drops they go on to complete that lap and the time counts. With the addition in this case that the pilot only has 30s to complete the lap. 

‘Overtime’ must, I think, have a different translation across the pond.

**Comment by HazardCreative** [comment]
Please don't misunderstand my tone. I don't intend to be combative, it's just that I honestly don't yet understand what specifically you cannot do.

**Comment by HazardCreative** [comment]
To run that format, I would:

1. Set up a "Practice" class and run practice through it. Use the data to set the lap count for mains. I would personally not delete the races, as they serve as seeding data for adaptive calibration.
2. Set up the main format something like this. You'll get a warning as having a lap count cap usually means you don't want a fixed time, but that won't stop you from doing it. I'm not sure if you have a standard for staging countdown, length, randomness, or number of tones; adjust accordingly.
![image](https://user-images.githubusercontent.com/7169527/121345007-580fbf80-c8f2-11eb-924e-260eab10e822.png)
3. Ensure that "Lap number" and "Time since start" are visible.
4. I agree that keeping track of the additional 30s cutoff is not ideal in the latest stable code. You can just roughly keep track with some other method and confirm laps afterward. (You can also use this branch to help, but it's definitely not considered stable code yet. https://github.com/RotorHazard/RotorHazard/tree/count-beyond-zero )
5. Racing stops and I have the following recorded:
![image](https://user-images.githubusercontent.com/7169527/121346900-a6be5900-c8f4-11eb-8f57-6d1369065d0c.png)
I take the following steps:
- R3 has a 7th lap. This happened after the end of the grace period at 2:31.8, so it can be deleted without question. Lap 6 is only 15s and the fastest of anyone by far; check in with pilots and spotters to see if it was valid and delete if necessary.
- R1 completed 6 laps in 1:55.4 and is the winner.
- The pilot on R6 has two highlights at the bottom of the list. Lap 5 completed at 2:01.5, so that pilot was already done. Delete lap 6.
- Everything looks in order for the pilot on R7.

I agree that enhancements could make this process smoother and simpler for race directors. But am I missing something in how the format is run that is outright not possible?

**Comment by Aaronsss** [comment]
Trying to use Rotorhazard in the UK does cause a lot of confusion with pilots as it doesn’t quite work the way of the established norm.

There is resistance by many to use Rotorhazard over LiveTime which is a shame. I think what’s described above is part of it but a lot of it is many race directors have invested in the lapRF8 and it has become the norm. Some still see it as Delta5 and aren't aware of how different Rotorhazard is now.

To me it sounds like #500 and #502 need to be implemented in combination with using the maximum lap counter? To get it run in the same manner as Livetime does currently?

While it could be run as @HazardCreative suggests that’s adds overheads for race directors and confusion to pilots as they won’t know when to land

@CerberusVelvet there is a schedule race option in race management tab that will count down until the next race but you do have to click the button manually after clicking save laps.

**Comment by HazardCreative** [comment]
Again—and please don't take this as putting up resistance—but what I'm looking for are the _specific behaviors_ needed to attain this.

What I think I'm hearing is that you'd like:
- #500 
- #502 
- #574
- #575
- #449 
- #501 
- #363
- Maybe #560
- Maybe #551 / #595
- Maybe #526 (hopefully already solved but I don't know that it's been tested)
- More data to be published about system accuracy and reliability

Does that cover it? What order would you prioritize them as?

**Comment by CerberusVelvet** [comment]
I think #500 is a definite priority for me. After running to the 'lap cap' format (bqe spec), yesterday. The thing that caused the most need for 'intervention' was pilots recording laps after the time had expired. If the laps stopped being counted, the race automatically moved to the stopped state, at the end of the timer. That would remove quite a lot of manual work for the RD.

**Comment by HazardCreative** [comment]
How do you propose handling false positives that inflate the lap count?

**Comment by CerberusVelvet** [comment]
In our use case we have two ‘stages’ of timer. The first timer is usually two minutes, but that is adjustable. After that timer elapses the second timer is thirty seconds, again that’s user configurable. After the first timer the pilots are told to finish the lap they are on. Once they record one more crossing no more crossings should be counted. Once the timer elapsed no more crossing should be counted. So if a false crossing occurs during the second timer, I would need the marshalling data to troubleshoot their last lap I guess. Any second crossing after the first crossing but before the timer has elapsed should probably be auto deleted in the same way as a crossing that’s below the minimum lap time? Then marshalled later?

An issue I had at the last race. I had two pilots battling for 2nd place. One did a fifth lap in 2:33 and one did a fifth lap in 2:40 (not actual data I’m rounding things up and down). On that circumstance the initial indication was that the pilot who did 5 in 2.33 was the winner. But actually neither of those 5th laps counted as they were after the time elapsed. So I had to look back for the time taken to complete 4 laps for each pilot and the other pilot had actually completed their fourth lap first. So they took the second place. I was able to figure it out. But then I had to jump through hoops to do that and then to prove it to the pilots and spectators who could see a different result on the stream overlays.

Once the second timer has elapsed then the race is done. No crossings after that timer count for anything. I’d be happy for the race status to move to stopped and anything after the timer expires to be disregarded in the results. Even if there had been a false crossing in the thirty seconds timer. It wouldn’t matter after the timer elapsed. Once time is up it’s all over. For us no crossings are relevant to the result after that timer has run out. 

Hope that makes some sort of sense!

**Comment by HazardCreative** [comment]
One thing you can do right now is set your race time to 2:00 instead of 2:30. This will help you notice right away when after-time-expired highlights appear is the lap list. If you see two, something is up! I will try to prioritize #574 so at minimum you will be able to see when the timer hits -30s.

Stopping after the grace period (#500) does make sense. A challenge there is that we cannot actually close and save timing immediately since crossings might have occurred before the time expired but are still in processing (the gate crossing has not yet cleared). We want to make sure that data gets added to the race appropriately if it came in before or after the buzzer. In this case, when you are inevitably asked "how much did I miss it by?" your only possible answer is "I don't know" since the automatic race termination will have and truncated that crossing forever. A pilot might argue they crossed inside of the timer's accuracy window; you would not be able to disprove this by showing the timing value is larger.

**Comment by HazardCreative** [comment]
Closing since this is simply a collection of separate behaviors and there's only one left which already has its own issue (#449).
